Add Navier-Stokes actions to simplify the input syntax

YaqiWang opened the pull request:
Close #15159.

YaqiWang committed:
merge Navier-Stokes actions into one and simplify the input syntax #15159

YaqiWang committed:
allow adding constant initial condition for vector variables when adding variables, for #15159

YaqiWang committed:
INS action #15159

YaqiWang commented:
Almost there, only Boussinesq approximation and RZ are not added.

YaqiWang commented:
A quick question @lindsayad : does hand-coded INS have Boussinesq approximation?

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/530727/) on 6e05655 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/15305/site)

This comment will be updated on new commits.

lindsayad commented:
> A quick question @lindsayad : does hand-coded INS have Boussinesq approximation?

No

YaqiWang commented:
Another one: AD does not have temperature time derivative kernel yet, right?

YaqiWang committed:
add temperature equation in INS action #15159

lindsayad commented:
No, but the heat conduction module does. The lid_driven/ad_lid_driven.i
test illustrates its use.

On Wed, May 20, 2020 at 10:22 AM Yaqi <notifications@github.com> wrote:

> Another one: AD does not have temperature time derivative kernel yet,
> right?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/15305#issuecomment-631613407>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACOGA4F32VHVZ3M2SG2QPJLRSQGU3ANCNFSM4NDFQVTA>
> .
>

YaqiWang committed:
add a parameter to optionally generate aux velocity vars when using AD #15159

YaqiWang commented:
Thanks @lindsayad. Just added Boussinesq approximation in action, will push it after lunch. I will not add RZ in this PR. I possibly will let my student add RZ support later. I will let you know when this is ready.

YaqiWang committed:
enable boussinesq approximation in action #15159

YaqiWang commented:
This is ready for review. You guys added lots of capabilities in INS, so the action is longer than what I thought ;-) Tag @lindsayad.

andrsd commented on modules/navier_stokes/doc/content/syntax/Modules/NavierStokes/index.md:
```diff
@@ -1 +1,15 @@
+
+!syntax list /Modules/NavierStokes objects=False actions=False subsystems=True
+
+!syntax list /Modules/NavierStokes objects=False actions=True subsystems=False
```
Can you remove:
```
- /Modules/NavierStokes
```
from `modules/navier_stokes/doc/hidden.yml`? Sinse you added doco, we need to remove it from the hidden file.

YaqiWang replied:
I am not quite familiar with MooseDocs, will take a more careful look.

andrsd replied:
we made the doco target/builds more draconian, so it will force you improve your moosedocs skills 🙃

lindsayad commented on modules/navier_stokes/src/actions/INSAction.C:
```diff
@@ -0,0 +1,928 @@
+    if (parameters.isParamSetByUser("convective_term"))
+      mooseWarning("'convective_term' is ignored for AD");
+    if (!getParam<bool>("laplace"))
+      mooseError("'laplace' must be true for AD");
```
Perhaps since you started working on this, we merged #15272, so AD now has support for the traction form

YaqiWang replied:
Looks like the main NS markdown file needs to be modified to reflect this. I was reading it and though traction form is not available for AD. I will remove this and update the index.md file.

lindsayad commented on modules/navier_stokes/src/actions/INSAction.C:
```diff
@@ -0,0 +1,928 @@
+  {
+    if (_use_ad)
+    {
+      if (getParam<bool>("supg"))
```
You also need `INSADTauMaterial` for PSPG

lindsayad commented on modules/navier_stokes/src/actions/INSAction.C:
```diff
@@ -0,0 +1,928 @@
+      {
+        const FunctionName funcx = _velocity_function[i * _dim];
+        if (funcx == "NA")
+          mooseError("AD does not support partial velocity BC");
```
`ADNodalBC` does now actually support partial specification of vector variable Dirichlet BCs via the `set_x_comp`, `set_y_comp`, and `set_z_comp` parameters. There is an example of using it in INS in `navier_stokes/test/tests/ins/RZ_cone/ad_rz_cone_by_parts.i` (and similar input files)

YaqiWang replied:
Good to know this!

lindsayad commented on modules/navier_stokes/src/postprocessors/NSEntropyError.C:
```diff
@@ -23,8 +23,8 @@ NSEntropyError::validParams()
   params.addRequiredParam<Real>("p_infty", "Freestream pressure");
-  params.addRequiredCoupledVar(NS::density, "density");
-  params.addRequiredCoupledVar(NS::pressure, "pressure");
+  params.addRequiredCoupledVar("rho", "density");
+  params.addRequiredCoupledVar("pressure", "pressure");
```
Why did this have to happen? Why couldn't these names still be available for the compressible Navier Stokes objects to use?

YaqiWang replied:
I guess this is the change from `NS::pressure` to `pressure`. I forgot the exact reason for changing this, possibly due to pressure naming with either `p` or `pressure`. I think using fixed parameter name is a better practice. Potentially `NS.h` can change.

lindsayad replied:
I feel like `NS.h` changing is not any more likely than someone coming along and trying to change the parameter name in `NSEntropyError.C` or other places. A nice thing about using something like `NS::pressure` everywhere is that if you ever did want to change the parameter name, you would only have to do it in one place. This isn't going to be the hill that I die on, but I would like to avoid having to regold things...

YaqiWang replied:
I think I remembered. I changed `NS.h` here for pressure name because INS is using `p` but here it has `pressure`. I guess NS.h was designed for CNS actions, so it is not an problem before. Now to be consistent, I choose to change the pressure name, which forces us to change lots of input files where `pressure` is used as the parameter name. This is the reason why I changed the parameter name not to use names from `NS` namespace. I think it is unlikely we change the parameter name in the future. Note that even we can change the parameter name in one place, it will affect the input files, which I can see will be a big pain. I double checked that the only regolding in this commit is due to the pressure variable name change.

lindsayad reviewed (changes requested):
This looks like very good work. A couple of questions:

1. Why did you completely revamp the compressible Navier Stokes system. If you have a good justification, and you are renaming a bunch of objects then I would pull for `Compressible` and `CNS` prefixes for consistency with the `INS` prefixes. Also I would like to see the `NS` names still be available for use in the compressible system
2. For the lid driven and some other tests, could you symlink to the non-action versions so we can just be doubly confident that they produce the same results? I didn't check to see whether you were using identical meshes or anything, but it would be nice to see some one-to-one correspondence.

YaqiWang commented:
Answer the two questions:
1. Renamed `NavierStokesAction` to `CNSAction` and  changed the syntax from `Modules/NavierStokes` to `Modules/CompressibleNavierStokes`.
2. Used the `file_base` in those input files with action so that I do not have to add new gold files.

YaqiWang commented:
Sorry, I had two force-pushes for addressing the comments. The first one is the main one. The second is for removing three gold exodus file under lid_driven test folder, that I forgot removing in the first push.

YaqiWang commented:
@lindsayad or @andrsd any more comments?

andrsd commented:
You seemed to fix the doco problems related to your PR, so that's fine with me. I will let @lindsayad comment on your code in case he has more to say.

YaqiWang commented:
It is strange that `Documentation` test is still failing. Maybe irrelevant with this PR?

lindsayad commented:
> It is strange that `Documentation` test is still failing. Maybe irrelevant with this PR?

correct, it is unrelated

lindsayad commented on modules/navier_stokes/test/tests/bump/bump.i:
```diff
@@ -111,32 +91,31 @@
     p_infty = 101325
     rho = rho
-    pressure = pressure
+    pressure = p
```
This is where I'm confused. In `NS.h` you changed the pressure variable name to be "p" instead of "pressure", which necessitates this change here, which I'm hoping is the only reason you had to re-gold the corresponding gold file. Why?

lindsayad replied:
You've manually changed the CNS object files to have the explicit parameter name "pressure", so why be inconsistent and change the input file variable name (generated by the action) to "p"?

It's not ideal having INS use a "p" parameter name in lots of places, while we are using "pressure" in CNS; that's definitely inconsistent. But it seems like you're adding additional inconsistency here instead of removing it...?

lindsayad commented on modules/navier_stokes/test/tests/ins/lid_driven/tests:
```diff
@@ -70,4 +70,42 @@
     requirement = 'We shall be able to solve the INS equations using the classical Chorin splitting algorithm'
   [../]
+
+  [./lid_driven_action]
+    type = 'Exodiff'
+    input = 'lid_driven_action.i'
+    exodiff = 'lid_driven_out.e'
+    custom_cmp = 'lid_driven.cmp'
+    prereq = 'lid_driven'
+    issues = '#15159'
+    requirement = 'We shall be able to reproduce the results from the input with MOOSE syntax only using the Modules/IncompressibleNavierStokes syntax'
+  [../]
+  [./lid_driven_stabilized_action]
+    type = 'Exodiff'
+    input = 'lid_driven_stabilized_action.i'
+    exodiff = 'lid_driven_stabilized_out.e'
+    custom_cmp = 'lid_driven_stabilized.cmp'
+    prereq = 'lid_driven_stabilized'
+    issues = '#15159'
+    requirement = 'We shall be able to reproduce the results from the input with MOOSE syntax only using the Modules/IncompressibleNavierStokes syntax'
+  [../]
+  [./ad_lid_driven_action]
+    type = 'Exodiff'
+    input = 'ad_lid_driven_action.i'
+    exodiff = 'lid_driven_out.e'
+    custom_cmp = 'lid_driven.cmp'
+    method = 'OPT'
+    prereq = 'ad_lid_driven'
+    issues = '#15159'
+    requirement = 'We shall be able to reproduce the results from the input with MOOSE syntax only using the Modules/IncompressibleNavierStokes syntax'
+  [../]
+  [./ad_lid_driven_stabilized_action]
+    type = 'Exodiff'
+    input = 'ad_lid_driven_stabilized_action.i'
+    exodiff = 'lid_driven_stabilized_out.e'
+    custom_cmp = 'lid_driven_stabilized.cmp'
+    prereq = 'ad_lid_driven_stabilized'
+    issues = '#15159'
+    requirement = 'We shall be able to reproduce the results from the input with MOOSE syntax only using the Modules/IncompressibleNavierStokes syntax'
+  [../]
```
This is great, thank you. Gives me the confidence that the results are the same!

The requirement phrasing is a little unclear to me. Moreover we are trying to make the requirement language be as separate as possible from the nature of the implementation. E.g. using an implementation name like `Modules/IncompressibleNavierStokes` is not ideal. Also, it looks like you've just used the same requirement four times :smile: 

I'd like to see something like: "The system shall be able to reproduce hand-coded Jacobian, unstabilized Navier-Stokes results using a customized and condensed action syntax." And then replace the first bit of that sentence depending on which requirement you're writing. Also, make sure you're adding periods to the end of requirements.

I'm only being strict about this because @permcody has hammered me on it many times.

YaqiWang replied:
Are you sure we need periods at the end? Looks like all other tests do not have period, ;-)

lindsayad replied:
Thanks for adding the periods to all!

lindsayad commented on modules/navier_stokes/include/base/NS.h:
```diff
@@ -19,10 +19,11 @@ const std::string momentum_y = "rhov";
 const std::string velocity_x = "vel_x";
 const std::string velocity_y = "vel_y";
 const std::string velocity_z = "vel_z";
-const std::string pressure = "pressure";
+const std::string pressure = "p";
```
Seems like if we undid this change here, then my concern in https://github.com/idaholab/moose/pull/15305#discussion_r428940253 would go away and we wouldn't have to re-gold.

YaqiWang replied:
Then INS action will be using pressure as the variable name and I will need to create more gold exodus files because of this.

YaqiWang replied:
Using p here is actually having a more consistent style as other names in NS.h.

YaqiWang commented:
The reason for that regolding is due to the variable name change from pressure to p but not the change in the postprocessor. I like ‘pressure’ as the parameter name and didn’t want to change the parameter name for both CNS and INS in this PR. Yes here I am only making the pressure variable name consistent between CNS and INS. Not at the side of my computer now, I will fix the requirement very soon.

lindsayad commented:
Ok, I'm satisfied. Just the requirements changes then

moosebuild commented:
All [jobs](https://civet.inl.gov/event/55139/) on dff48f0 : invalidated by @lindsayad

come on CIVET

lindsayad commented:
Ugh CIVET appears not to be getting payloads for this new commit...

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/532169/) on dff48f0 : invalidated by @lindsayad

lindsayad commented:
@YaqiWang Can you rebase on devel and force push? We need something to retrigger testing

lindsayad reviewed (approved):
Thanks a lot for this large contribution @YaqiWang !
